### PR TITLE
[Perf] Stacked audio ops: batched matmul and embedding gather in MOSS-TTS talker

### PR DESCRIPTION
**Pull request by yangyonggit**
## Purpose

Replace per-quantizer Python loops in the MOSS-TTS talker with batched tensor ops, reducing CPU overhead on every decode step.

The delay-model talker (`MossTTSDelayTalkerForGeneration`) and the realtime talker (`MossTTSRealtimeTalkerForGeneration`) both iterate over `n_vq` quantizer heads at each decode step — once for audio embedding lookup and once for audio head logits (delay variant only). At `n_vq=16` or `n_vq=32` this creates a tight Python loop that serialises what are otherwise independent tensor ops.

**Changes:**

After `load_weights()`, both talkers pre-stack their per-quantizer weights into a single tensor:
- `_stacked_audio_emb_w`: `(n_vq, vocab+1, hidden_size)` — used for a single batched gather instead of n_vq embedding lookups
- `_stacked_audio_head_w` (delay talker only): `(n_vq, vocab+1, hidden_size)` — used for a single batched matmul instead of n_vq serial `nn.Linear` calls

The original loop is kept as a fallback path (used before `load_weights()` completes).

> **Note:** The E2E tests in this PR depend on #4220. Please merge #4220 before merging this PR.

## Test Plan

MOSS-TTS E2E tests covering both the delay-model talker (VoiceGenerator) and the realtime talker:

```bash
pytest tests/e2e/offline_inference/test_moss_tts.py -v
pytest tests/e2e/offline_inference/test_moss_tts_realtime.py -v


## Test Result

tests/e2e/offline_inference/test_moss_tts.py
================== 3 passed, 18 warnings in 153.18s ==================

tests/e2e/offline_inference/test_moss_tts_realtime.py
================== 2 passed, 18 warnings in 164.25s ==================

## log
[stack_e2e.log](https://github.com/user-attachments/files/28670985/stack_e2e.log)
[stack_e2e_realtime.log](https://github.com/user-attachments/files/28670986/stack_e2e_realtime.log)

### CODE REVIEW

**Review by hsliuustc0106 — COMMENTED** [review]
**BLOCKER scan:**

- **Correctness** — The stacked op paths look correct. Fallback logic preserved for pre-load_weights().
- **Reliability** — The null checks on `_stacked_audio_*` guards are correct.

**Non-blocking:**

1. ~vllm_omni/model_executor/models/moss_tts/modeling_moss_tts_talker.py:424~ — The `unsqueeze(1)` creates shape `(n_vq, 1, hidden_size)` then indexing with `codes.t()` of shape `(L, n_vq)` → advanced indexing produces `(n_vq, L, hidden_size)`. This works but is hard to follow. Consider pre-transposing codes to `(n_vq, L)` earlier to make the intent explicit.

2. The test rewrite in test_moss_tts.py is a good cleanup — AutoProcessor integration matches the upstream examples/offline_inference pattern.

3. ~test_moss_tts.py:180~ — The assertion `assert len(results) >= 2` for batch test is correct for async_chunk streaming, but the comment \"we need at least one non-empty audio tensor per request\" might mislead future readers. With FINAL_ONLY consolidation, each request yields exactly one audio tensor. Consider clarifying: \"With async_chunk streaming + FINAL_ONLY, each request produces one consolidated output. The >= 2 guard handles backward-compatibility if consolidation behavior changes.\"

VERDICT: COMMENT

**Review by linyueqian — COMMENTED** [review]
The embedding-gather replacements look numerically exact: the stacked gather plus sum-over-quantizers matches the per-head loop, the clamp ranges line up with the embedding dimensions, and batch-size-1 and concurrent ordering are preserved. The audio-head matmul (`stacked_w @ h`) is also equivalent to the per-head `head(h)` row-wise.

One equivalence concern I want to flag before this lands:

[HIGH] In the delay-talker batched sampling path, the new code inserts `all_logits = all_logits.float()` before temperature/top-k/softmax/multinomial. The original per-head path (`_sample_with_top_k`) runs softmax and multinomial in the model dtype (bf16/fp16 in normal serving), with no float cast. softmax+multinomial in fp32 vs bf16 produce different probability vectors, so for the same RNG state the sampled codes diverge. This is a behavior change, not a bit-exact refactor. If the intent is "same output, just faster," drop the `.float()` so the math stays in model dtype. If fp32 sampling is actually desired (arguably more correct), please call it out explicitly in the PR description and note that audio output will differ from the pre-PR baseline.

[LOW] The PR removes the old `torch.allclose` determinism assertion and adds no test comparing the stacked path against the loop fallback, so this divergence would not be caught. A small unit test toggling the stacked-weight buffers to None and asserting `allclose` on embeds and pre-sample logits would lock in equivalence.

@gcanlin would value your read on whether the fp32 sampling change is intended for MOSS-TTS. I verified the dtype divergence from the code but did not measure the audio-quality delta.
